### PR TITLE
GODRIVER-3061 Expand the capabilities of the docker container

### PR DESCRIPTION
**Pull request by blink1073**
Allows the docker container to be split into server and client components, so that other services like the load balancer can be run in parallel.

Tested with https://github.com/mongodb/mongo-go-driver/pull/1494

### CODE REVIEW

**Review by prestonvasquez — COMMENTED** [review]
When I run locally using this script:

```bash
cd ${DRIVERS_TOOLS}/.evergreen/docker
bash ./run-server.sh \
```

I get the following error:

```
❯ ./run-server.sh
[+] Building 67.6s (10/10) FINISHED                                                                                                                                 docker:desktop-linux
 => [internal] load build definition from Dockerfile                                                                                                                                0.0s
 => => transferring dockerfile: 937B                                                                                                                                                0.0s
 => [internal] load .dockerignore                                                                                                                                                   0.0s
 => => transferring context: 2B                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/ubuntu:20.04                                                                                                                     1.9s
 => [1/5] FROM docker.io/library/ubuntu:20.04@sha256:f5c3e53367f142fab0b49908550bdcdc4fb619d2f61ec1dfa60d26e0d59ac9e7                                                               0.0s
 => => resolve docker.io/library/ubuntu:20.04@sha256:f5c3e53367f142fab0b49908550bdcdc4fb619d2f61ec1dfa60d26e0d59ac9e7                                                               0.0s
 => => sha256:51abb690a3439f2dc40cd5546d184a849ecba847491d2acbaebaec01d90196db 2.32kB / 2.32kB                                                                                      0.0s
 => => sha256:f5c3e53367f142fab0b49908550bdcdc4fb619d2f61ec1dfa60d26e0d59ac9e7 1.13kB / 1.13kB                                                                                      0.0s
 => => sha256:074f14759aac23ee1f3ec318fbc1baa83f309ff78ad4448d63f79d2c8bea24c3 424B / 424B                                                                                          0.0s
 => [internal] load build context                                                                                                                                                   0.0s
 => => transferring context: 116B                                                                                                                                                   0.0s
 => [2/5] RUN export DEBIAN_FRONTEND=noninteractive &&   apt-get -qq update && apt-get -qqy install --no-install-recommends   git   ca-certificates   curl   wget   sudo   gnupg   64.9s
 => [3/5] COPY ./local-entrypoint.sh /root/local-entrypoint.sh                                                                                                                      0.0s
 => [4/5] COPY ./base-entrypoint.sh /root/base-entrypoint.sh                                                                                                                        0.0s
 => [5/5] COPY ./test-entrypoint.sh /root/test-entrypoint.sh                                                                                                                        0.0s
 => exporting to image                                                                                                                                                              0.7s
 => => exporting layers                                                                                                                                                             0.7s
 => => writing image sha256:318efc209e4afda2ecb72bbadf3fab0377634f993d993c549da6253abe072373                                                                                        0.0s
 => => naming to docker.io/library/drivers-evergreen-tools                                                                                                                          0.0s

What's Next?
  1. Sign in to your Docker account → docker login
  2. View a summary of image vulnerabilities and recommendations → docker scout quickview
Finding Python3 binary...
Finding Python3 binary... done.
linux-ubuntu-20.04-aarch64
http://downloads.10gen.com/linux/mongodb-linux-aarch64-enterprise-ubuntu2004-latest.tgz
Installing server binaries...
Installing server binaries... done.
mv: cannot move 'mongodb' to a subdirectory of itself, 'mongodb/mongodb'
mv: cannot move 'mongodb-linux-aarch64-enterprise-ubuntu2004-7.3.0-alpha-864-g8ff75de' to 'mongodb/mongodb-linux-aarch64-enterprise-ubuntu2004-7.3.0-alpha-864-g8ff75de': Directory not empty
```
Do we need [L692](https://github.com/mongodb-labs/drivers-evergreen-tools/blob/7f42d84ab09ab071136ce46ede9d79fd113922f9/.evergreen/download-mongodb.sh#L692) of the download-mongodb.sh script? Removing it resolves this problem.

---

Is there a way to use non-server specific orchestrations, such a replica set orchestrations?

```
cd ${DRIVERS_TOOLS}/.evergreen/docker
ORCHESTRATION_FILE="replica_sets/basic.json" \
  bash ./run-server.sh 
```

**Comment by blink1073** [comment]
> Do we need [L692](https://github.com/mongodb-labs/drivers-evergreen-tools/blob/7f42d84ab09ab071136ce46ede9d79fd113922f9/.evergreen/download-mongodb.sh#L692) of the download-mongodb.sh script? Removing it resolves this problem.

Fixed by removing the `mongodb` directory, in case we are switching versions.

File: .evergreen/docker/run-client.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+PLATFORM=${DOCKER_PLATFORM:-}
+# e.g. --platform linux/amd64
+
+# Mongo orchestration related variables.
+MONGODB_VERSION=${MONGODB_VERSION:-latest}
+TOPOLOGY=${TOPOLOGY:-replica_set}
+ORCHESTRATION_FILE=${ORCHESTRATION_FILE:-basic.json}
+SKIP_CRYPT_SHARED_LIB=${SKIP_CRYPT_SHARED_LIB:-false}
+AUTH=${AUTH:-""}
+SSL=${SSL:=""}
```
[the remark this answers is not in the document]
Comment (blink1073) [reply .evergreen/docker/run-client.sh]: Fixed

File: .evergreen/docker/run-client.sh [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+# Build up the arguments.
+ARGS="$PLATFORM --rm -i"
+ARGS+=" -e MONGODB_VERSION=$MONGODB_VERSION -e TOPOLOGY=$TOPOLOGY"
+ARGS+=" -e SSL=$SSL -e AUTH=$AUTH"
+ARGS+=" -e MONGODB_BINARIES=$MONGODB_BINARIES"
+ARGS+=" -e CRYPT_SHARED_LIB_PATH=$MONGODB_BINARIES/mongosh_crypt_v1.so"
+ARGS+=" -e ORCHESTRATION_FILE=$ORCHESTRATION_FILE"
+ARGS+=" -e SKIP_CRYPT_SHARED_LIB=$SKIP_CRYPT_SHARED_LIB"
+ARGS+=" -e DRIVERS_TOOLS="
```
[the remark this answers is not in the document]
Comment (blink1073) [reply .evergreen/docker/run-client.sh]: Fixed

File: .evergreen/docker/README.md [hunk 3]
```diff
@@ -22,22 +22,59 @@ will be passed through to the container.
...
+TOPOLOGY
+LOAD_BALANCER
+STORAGE_ENGINE
+REQUIRE_API_VERSION
+DISABLE_TEST_COMMANDS
+MONGODB_VERSION
+MONGODB_DOWNLOAD_URL
+ORCHESTRATION_FILE
+```
+
```
Comment (prestonvasquez) [inline .evergreen/docker/README.md:42]: Suggest adding the following explanation on the default behavior of TOPOLOGY and ORCHESTRATION_FILE:

````suggestion

Note that the default `TOPOLOGY` is [`servers`](https://github.com/mongodb-labs/drivers-evergreen-tools/tree/master/.evergreen/orchestration/configs/servers) and the default `ORCHESTRATION_FILE` is `basic.json`. For example, to run a replica using the [auth](https://github.com/mongodb-labs/drivers-evergreen-tools/blob/master/.evergreen/orchestration/configs/replica_sets/auth.json) file orchestration:

```bash
TOPOLOGY=replica_set ORCHESTRATION_FILE=auth.json bash ./run-server.sh 
```
````

**Review by prestonvasquez — APPROVED** [review]

**Pull request merged**